File: core/mycelium.py
"""
Mycium AI Framework — Core Orchestration (implemented)
"""

import yaml
import os
from pathlib import Path
from typing import List, Optional
from rhizomorph.lcm import Rhizomorph
from core.providers import AgentProvider, MockAgentProvider

FRAMEWORK_ROOT = Path(__file__).parent.parent


class Mycelium:
    """
    The brain. Reasons, routes, delegates. Never executes. Never sees.
    """

    def __init__(self, provider: Optional[AgentProvider] = None, rhizomorph_db: str = "./lcm.db"):
        self.provider = provider or MockAgentProvider()
        self.models = self._load_config("config/models.yaml")
        self.rules = self._load_config("config/rules.yaml")
        self.registry = self._load_registry()
        self.rhizo = Rhizomorph(rhizomorph_db)

    def _load_config(self, path):
        with open(FRAMEWORK_ROOT / path) as f:
            return yaml.safe_load(f)

    def _load_registry(self):
        roles = {}
        roles_dir = FRAMEWORK_ROOT / "registry" / "roles"
        # Load roles from subdirectories (e.g., design/, marketing/)
        for dept_dir in roles_dir.iterdir():
            if dept_dir.is_dir():
                dept = dept_dir.name
                roles[dept] = []
                for role_file in dept_dir.glob("*.md"):
                    roles[dept].append(role_file.stem)
        # Also load roles directly in roles/ (e.g., engineering-*.md)
        root_roles = []
        for role_file in roles_dir.glob("*.md"):
            root_roles.append(role_file.stem)
        if root_roles:
            roles["engineering"] = root_roles
        return roles

    def get_model_for_role(self, role):
        return self.models["roles"].get(role, {}).get("id")

    def get_model_for_task(self, task_type):
        matrix = self.models.get("task_matrix", {})
        task = matrix.get(task_type, matrix.get("research"))
        return task["primary"] if task else self.models["roles"]["dynamic"]["id"]

    def find_agent_role(self, capability):
        for dept, roles in self.registry.items():
            for role in roles:
                if capability.lower() in role.lower():
                    return {"department": dept, "role": role}
        return None

    def check_colony_health(self):
        """Query Rhizomorph for mission-related events and summarize."""
        health = {"status": "healthy", "active_missions": 0, "completed_missions": 0, "pain_points": 0, "green_leaves": 0, "recent_benchmarks": 0}
        # count events by tag
        missions = self.rhizo.get_events_by_tag("#mission", limit=100)
        completed = self.rhizo.get_events_by_tag("#mission-complete", limit=100)
        pain = self.rhizo.get_events_by_tag("#pain-point", limit=100)
        green = self.rhizo.get_events_by_tag("#green-leaf", limit=100)
        bench = self.rhizo.get_events_by_tag("#benchmark", limit=100)

        health.update({
            "active_missions": len(missions),
            "completed_missions": len(completed),
            "pain_points": len(pain),
            "green_leaves": len(green),
            "recent_benchmarks": len(bench),
        })
        return health

    def execute_mission(self, mission: str, capabilities: List[str] = None):
        """Create a mission event, delegate to ArmyAnt, return dispatch summary."""
        capabilities = capabilities or ["frontend", "design"]
        # write mission start to rhizomorph
        seq = self.rhizo.emit("memory.write", {"mission": mission}, agent="mycelium", tags=["#mission"])
        army = ArmyAnt(self, provider=self.provider)
        team = army.build_team(mission, capabilities)
        dispatch = army.coordinate(team["team"], mission)
        return {"mission": mission, "rhizo_seq": seq, "dispatch": dispatch}


class Scout:
    def __init__(self, mycelium: Mycelium):
        self.mycelium = mycelium
        self.model = mycelium.get_model_for_role("scout")
        self.provider = mycelium.provider
        self.rhizo = mycelium.rhizo

    def research_models(self, query: str):
        task = f"research free models for: {query}"
        resp = self.provider.spawn_agent(self.model, task)
        # write result to rhizomorph
        tags = resp.get("tags", ["#benchmark"]) if isinstance(resp, dict) else ["#benchmark"]
        self.rhizo.emit("memory.write", {"query": query, "result": resp}, agent="scout", tags=tags)
        return resp

    def find_opportunities(self, domain: str):
        task = f"opportunity scan: {domain}"
        resp = self.provider.spawn_agent(self.model, task)
        tags = resp.get("tags", ["#green-leaf"]) if isinstance(resp, dict) else ["#green-leaf"]
        self.rhizo.emit("memory.write", {"domain": domain, "result": resp}, agent="scout", tags=tags)
        return resp


class ArmyAnt:
    def __init__(self, mycelium: Mycelium, provider: Optional[AgentProvider] = None):
        self.mycelium = mycelium
        self.provider = provider or mycelium.provider
        self.model = mycelium.get_model_for_role("army")
        self.rhizo = mycelium.rhizo

    def build_team(self, mission: str, required_capabilities: List[str]):
        team = []
        for cap in required_capabilities:
            role = self.mycelium.find_agent_role(cap)
            model = self.mycelium.get_model_for_task(cap)
            team.append({"role": role, "model": model, "capability": cap})
        # write team plan to rhizomorph
        self.rhizo.emit("memory.write", {"mission": mission, "team": team}, agent="army", tags=["#mission"])
        return {"team": team}

    def coordinate(self, team: List[dict], mission: str):
        results = []
        for member in team:
            role = member.get("role") or {"role": "unknown"}
            model = member.get("model")
            cap = member.get("capability")
            task = f"Execute {cap} for mission: {mission}"
            ant = DynamicAnt(role, model, task, provider=self.provider, rhizo=self.rhizo)
            res = ant.execute()
            results.append(res)
        # emit coordination complete
        self.rhizo.emit("memory.write", {"mission": mission, "coordinated": True}, agent="army", tags=["#mission"])
        return results


class DynamicAnt:
    def __init__(self, role, model, task, provider: Optional[AgentProvider] = None, rhizo: Optional[Rhizomorph] = None):
        self.role = role
        self.model = model
        self.task = task
        self.provider = provider or MockAgentProvider()
        self.rhizo = rhizo or Rhizomorph()

    def execute(self):
        # write mission start
        self.rhizo.emit("memory.write", {"task": self.task, "role": self.role}, agent=str(self.role), tags=["#mission"])
        resp = self.provider.spawn_agent(self.model, self.task)
        # if provider returned error, write pain-point
        tags = resp.get("tags", []) if isinstance(resp, dict) else []
        if isinstance(resp, dict) and resp.get("result", {}).get("error"):
            self.rhizo.emit("memory.write", {"task": self.task, "error": resp}, agent=str(self.role), tags=["#pain-point"])
            return {"role": self.role, "status": "failed", "resp": resp}

        # normal completion
        self.rhizo.emit("memory.write", {"task": self.task, "result": resp}, agent=str(self.role), tags=["#mission-complete"])
        return {"role": self.role, "status": "completed", "resp": resp}


# CLI entrypoint support
if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    sub = parser.add_subparsers(dest="cmd")
    m = sub.add_parser("mission")
    m.add_argument("mission")
    m.add_argument("--provider", default="mock")
    sub.add_parser("health")
    s = sub.add_parser("scout")
    s.add_argument("query")

    args = parser.parse_args()
    provider = MockAgentProvider() if getattr(args, "provider", "mock") == "mock" else None
    brain = Mycelium(provider=provider)
    if args.cmd == "mission":
        print(brain.execute_mission(args.mission))
    elif args.cmd == "health":
        print(brain.check_colony_health())
    elif args.cmd == "scout":
        sc = Scout(brain)
        print(sc.research_models(args.query))

File: src/mycelium.mjs
/**
 * Mycelium — Colonial AI Agent Orchestrator
 * Thinks, plans, dispatches. NEVER executes leaf work.
 * Every action is an event in Rhizomorph.
 */
import { Rhizomorph } from './rhizomorph.mjs';
import { createProvider } from './providers.mjs';
import { readdirSync, readFileSync, existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { parse as parseYaml } from './yaml.mjs';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, '..');

export class ModelConfig {
  constructor(configPath = join(ROOT, 'config', 'models.yaml')) {
    this.roles = {};
    this.fallbacks = [];
    if (!existsSync(configPath)) return;
    const cfg = parseYaml(readFileSync(configPath, 'utf8'));
    if (cfg.roles) this.roles = cfg.roles;
    if (cfg.fallbacks) this.fallbacks = Array.isArray(cfg.fallbacks) ? cfg.fallbacks : [];
  }
  modelFor(role) {
    const entry = this.roles[role];
    if (entry?.id && entry?.provider) return `${entry.provider}/${entry.id}`;
    if (entry?.id) return entry.id;
    return this.roles.mycelium?.id ? `kilocode/${this.roles.mycelium.id}` : 'kilocode/xiaomi/mimo-v2-pro:free';
  }
}

export class RoleRegistry {
  constructor(rolesDir = join(ROOT, 'registry', 'roles')) {
    this.roles = new Map();
    if (!existsSync(rolesDir)) return;
    for (const entry of readdirSync(rolesDir, { withFileTypes: true })) {
      if (entry.isDirectory()) {
        const dept = entry.name;
        for (const file of readdirSync(join(rolesDir, dept)).filter(f => f.endsWith('.md'))) {
          this.roles.set(file.replace('.md', ''), { department: dept, role: file.replace('.md', '') });
        }
      } else if (entry.isFile() && entry.name.endsWith('.md')) {
        this.roles.set(entry.name.replace('.md', ''), { department: 'general', role: entry.name.replace('.md', '') });
      }
    }
  }
  find(cap) { for (const [r, info] of this.roles) if (r.toLowerCase().includes(cap.toLowerCase())) return info; return null; }
  list() { return Array.from(this.roles.values()); }
  count() { return this.roles.size; }
}

export class DynamicAnt {
  constructor(provider, rhizo, modelConfig) {
    this.provider = provider;
    this.rhizo = rhizo;
    this.models = modelConfig || new ModelConfig();
  }

  _gatherContext(role) {
    const lessons = this.rhizo.query({ tags: ['#lesson'], limit: 5 });
    const painPoints = this.rhizo.query({ tags: ['#pain-point'], limit: 3 });
    const shortcuts = this.rhizo.query({ tags: ['#shortcut'], limit: 3 });
    let ctx = '';
    if (lessons.length) ctx += `\n## Colony Lessons\n${lessons.map(e => `- ${typeof e.payload === 'string' ? e.payload : JSON.stringify(e.payload)}`).join('\n')}`;
    if (painPoints.length) ctx += `\n## Known Pain Points\n${painPoints.map(e => `- ${typeof e.payload === 'string' ? e.payload : JSON.stringify(e.payload)}`).join('\n')}`;
    if (shortcuts.length) ctx += `\n## Shortcuts\n${shortcuts.map(e => `- ${typeof e.payload === 'string' ? e.payload : JSON.stringify(e.payload)}`).join('\n')}`;
    return ctx;
  }

  _selfEval(task, role, result) {
    const hasError = result.error;
    const scores = {
      accuracy: hasError ? 2 : 4,
      efficiency: hasError ? 2 : 4,
      completeness: hasError ? 1 : 4,
      reusability: hasError ? 1 : 3,
    };
    const avg = (scores.accuracy + scores.efficiency + scores.completeness + scores.reusability) / 4;
    const tags = [];
    if (avg >= 4) tags.push('#lesson');
    if (scores.efficiency < 4) tags.push('#shortcut');
    if (avg < 3) tags.push('#pain-point');
    tags.push('#self-eval');
    this.rhizo.emit('agent.self-eval', { task, role, scores, average: avg }, { agent: `dynamic-${role}`, tags });
    return scores;
  }

  execute({ task, role, model }) {
    const context = this._gatherContext(role);
    this.rhizo.emit('memory.write', { task, role }, { agent: 'dynamic-ant', tags: ['#mission'] });
    const enrichedTask = context
      ? `Registry role: ${role}\n\n${context}\n\nTASK: ${task}`
      : `Registry role: ${role}\nTASK: ${task}`;
    const result = this.provider.spawnAgent({
      model: model || this.models.modelFor('dynamic'),
      task: enrichedTask,
      agentId: `dynamic-${role}`,
    });
    this.rhizo.emit('memory.write', { task, role, result: result.result }, {
      agent: 'dynamic-ant',
      tags: result.error ? ['#pain-point'] : ['#mission-complete'],
    });
    const scores = this._selfEval(task, role, result);
    return { role, status: result.error ? 'failed' : 'completed', resp: result, scores };
  }
}

export class ArmyAnt {
  constructor(provider, rhizo, registry, modelConfig) {
    this.provider = provider;
    this.rhizo = rhizo;
    this.registry = registry;
    this.models = modelConfig || new ModelConfig();
  }
  buildTeam(mission, capabilities) {
    const team = capabilities.map(cap => ({
      capability: cap,
      role: this.registry.find(cap) || { department: 'general', role: `general-${cap}` },
      model: this.models.modelFor('dynamic'),
    }));
    this.rhizo.emit('memory.write', { mission, team }, { agent: 'army-ant', tags: ['#mission'] });
    return { mission, team };
  }
  coordinate(team, task) {
    const ant = new DynamicAnt(this.provider, this.rhizo, this.models);
    return team.map(m => ant.execute({ task, role: m.role?.role || m.capability, model: m.model }));
  }
}

export class Scout {
  constructor(provider, rhizo, modelConfig) {
    this.provider = provider;
    this.rhizo = rhizo;
    this.models = modelConfig || new ModelConfig();
  }

  _priorBenchmarks(domain) {
    const benchmarks = this.rhizo.query({ tags: ['#benchmark'], limit: 5 });
    if (!benchmarks.length) return '';
    return `\n## Prior Benchmarks\n${benchmarks.map(e => `- ${typeof e.payload === 'string' ? e.payload : JSON.stringify(e.payload)}`).join('\n')}\n`;
  }

  research(query) {
    this.rhizo.emit('memory.write', { query }, { agent: 'scout', tags: ['#mission'] });
    const prior = this._priorBenchmarks(query);
    const enrichedQuery = prior
      ? `SCOUT. Research only, never execute.\n${prior}\n${query}`
      : `SCOUT. Research only, never execute.\n${query}`;
    const result = this.provider.spawnAgent({
      model: this.models.modelFor('scout'),
      task: enrichedQuery,
      agentId: 'scout',
    });
    this.rhizo.emit('memory.write', { query, result: result.result }, { agent: 'scout', tags: ['#benchmark'] });
    return { query, ...result };
  }

  findOpportunities(domain) {
    const result = this.provider.spawnAgent({
      model: this.models.modelFor('scout'),
      task: `SCOUT. Find revenue opportunities in: ${domain}`,
      agentId: 'scout-opp',
    });
    this.rhizo.emit('memory.write', { domain, result: result.result }, { agent: 'scout', tags: ['#green-leaf'] });
    return { domain, ...result };
  }
}

export class Mycelium {
  constructor(opts = {}) {
    this.provider = opts.provider || createProvider(opts.providerType || 'auto');
    this.rhizo = new Rhizomorph(opts.dbPath || join(ROOT, 'lcm.db'));
    this.registry = new RoleRegistry();
    this.models = new ModelConfig(opts.modelsPath);
  }

  executeMission(mission, capabilities = ['frontend']) {
    this.rhizo.emit('memory.write', { mission }, { agent: 'mycelium', tags: ['#mission'] });
    const army = new ArmyAnt(this.provider, this.rhizo, this.registry, this.models);
    const plan = army.buildTeam(mission, capabilities);
    const results = army.coordinate(plan.team, mission);
    return { mission, rhizoSeq: this.rhizo.seq, dispatch: results };
  }

  checkColonyHealth() {
    const missions = this.rhizo.countByTag('#mission');
    const completed = this.rhizo.countByTag('#mission-complete');
    const painPoints = this.rhizo.countByTag('#pain-point');
    const greenLeaves = this.rhizo.countByTag('#green-leaf');
    const benchmarks = this.rhizo.countByTag('#benchmark');
    const lessons = this.rhizo.countByTag('#lesson');
    const selfEvals = this.rhizo.countByTag('#self-eval');
    return {
      status: painPoints > completed ? 'unhealthy' : 'healthy',
      activeMissions: missions - completed,
      completedMissions: completed,
      painPoints,
      greenLeaves,
      recentBenchmarks: benchmarks,
      lessons,
      selfEvals,
    };
  }

  scout() { return new Scout(this.provider, this.rhizo, this.models); }
  army() { return new ArmyAnt(this.provider, this.rhizo, this.registry, this.models); }
  close() { this.rhizo.close(); }
}
